### ISSUE TITLE
Which GPU is supported for simultaneous graphical desktop usage?

### ISSUE THREAD

**Issue by seesturm** [ROCmSoftwarePlatform/tensorflow-upstream#312]
As reported in a separate issue my RX 460 has troubles running tensorflow with some kinds of background load. While it might be acceptable to have a requirement of not saturating the CPU with workload, it is not practical for me to set aside the whole desktop PC just for machine learning. Even for doing machine learning alone it is needed to do other stuff like reading web documentation in parallel.

But it appears that for a bug report to be acceptable one has to declare that the whole PC is just used for the very ML task and nothing else. The question is now, is there a GPU for which mild simultaneous desktop usage is supported? Are bug reports are accepted for such a use-case?

I'm considering buying a more powerful GPU (probably Vega) but need this information for a buying decision.

**Comment by sunway513** [comment]
Hi @seesturm , Tensorflow training by default will allocate all the available device memories upon initialization; therefore, we recommend users keep the GPU dedicated for TF applications as possible.
For the issue you've cited (#300), the RX460 GPU only has 2GB memory, which is not sufficient for moderate TF applications alone. 
I'd like @jlgreathouse to help weigh in your request on GPU selection.

**Comment by jlgreathouse** [comment]
I can't really make an educated comment here. While I work on a bunch of software in the ROCm stack, I don't personally spend a lot of time running our TF framework. Additionally, because I have to juggle dozens of GPUs to handle all the random questions and bug reports we get in the ROCm issue trackers, I do not really use any of my ROCm test systems as a GUI-driven desktop PC..

**Comment by Burstaholic** [comment]
This may be helpful - there are a couple of ways to control VRAM usage, either a fixed percentage, or just forcing it to allocate lazily:

https://stackoverflow.com/questions/34199233/how-to-prevent-tensorflow-from-allocating-the-totality-of-a-gpu-memory

**Comment by seesturm** [comment]
Hi @Burstaholic. Thanks for the pointer. I tried this right now, but am still running into the same trouble. After some random time tensorflow freezes or runs into a memory access error.

**Comment by sunway513** [comment]
Hi @seesturm , for RC460, to play around the TF memory allocation methods might not help; especially if there're several processes competing for the limited GPU computing resources and memories.

Could you elaborate on the "freeze"? Is that a system hard hang (require a system reset to get back working) or just the graphics becoming slow and unresponsive for a while, but can recover later without a system reset?
As I've confirmed with our amdgpu kernel experts, the latter situation can be normal, as compute and graphics would compete for the same shader resources. 
In the driver level, we already have queue priorities that should give graphics a higher priority than compute. 
However, it might still be possible to have situations where certain compute workloads can starve graphics for resources.

Like what we've mentioned on issue #300  , the cifar10 training was working correctly on my RX480 workstation with the cited workaround applied for gfx803 target.

**Comment by seesturm** [comment]
Hi @sunway513. I did a inspection of the cifar10 model referenced in #300 and it looks like as it uses very little memory (it is a model with less than 10 layers). So memory starvation is indeed very unlikely.

The freeze only concerns the tensorflow process itself. When the freeze happens rocm-smi reports 100% GPU usage and 20W power consumption. (Before freezing the consumption is 35-45W). The tensorflow process (python3) consumes about 9% of a single CPU core during the freeze. Tensorflow cannot be terminated using CTRL-C, but sending a SIGQUIT does the trick.

While tensorflow is frozen the rest of the system has no observable degradation. The system is 100% responsive and video playback shows no degradation at all. I can even launch 3D Graphics (game) and while it shows lower performance (about 1/2) there is no stutter. The only time when the rest of the system was severely affected was the HIP test 74 mentioned in my [issue 300 comment](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/300#issuecomment-457909607). But after the failed test finished the system did completely recover.

Another recent finding is that I can rather instantly trigger the memory access error (for cifar10_train.py) during remote access (outside of graphical session) by means of executing the cifar10_eval.py in parallel to the training (it us supposed to be used that way). So I'm starting to suspect that this is not a desktop vs non-desktop issue.

**Comment by sunway513** [comment]
Hi @seesturm , thanks for the clarifications. From what you've described here, your issue was soft hang, which can fall into the situation I've mentioned in the last comment (competes for shader resources).
When there're multiple processes using the limited RX460 GPU resources, the situation can become very complicated down to the system kernel driver level. 
Some tricks might help:
1. To suspend tensorflow process, you can use CTRL-Z.
2. The following script can check the realtime GPU memory utilization:
```
for file in `find /sys/class/kfd/kfd/topology/nodes/ -name used_memory`
do
  mem=`cat $file`
  step=1024
  mem=`echo "scale=2; $mem/$step/$step" | bc`
  echo $file ${mem}MB
done
```

**Comment by seesturm** [comment]
I want to point out that with parallel execution of "cifar10_eval.py" I'm not experiencing a "hang" but an immediate tensorflow termination due to "Memory access fault by GPU node-2".

Also tried to obtain the GPU memory utilization. But the "used_memory" sysfs interface seems not to be available with kernel 4.20.6.

**Comment by sunway513** [comment]
Hi @seesturm , I'm in progress reproducing the issue you mentioned on "cifar10_eval.py".
Previously we don't have any use case to run multiple tensorflow instances on the same GPU target.

**Comment by sunway513** [comment]
Hi @seesturm , the failure you mentioned should still be related to the small device memory on RX460.
After applying the tricks mentioned [above](https://stackoverflow.com/questions/34199233/how-to-prevent-tensorflow-from-allocating-the-totality-of-a-gpu-memory), I can limit each of the tensorflow instances only allocate 1/3 of the device memory available. 

The following screenshot was collected on my test node with RX480 4GB:
https://www.dropbox.com/s/4oycwxteyk3osdc/tf-1.12-cifar10.png
Left window was running the [cifar10_multigpu_train.py](https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py), the right upper windows shows the device memory footprint with two TF instances, which is slightly more than 3 GB.
Both of the instances run correctly (as shown in the screenshot), and I didn't see any crashes or vmem fault.

In addition, if I connect the RX480 card to my 4k monitor, the graphic rendering will take 700MB+ device memories.
So back to the question on this issue, as a basic suggestion, you should pick up a ROCm supported GPU with sufficient size of device memory, e.g. minimum 4GP for cifar10 sample, if you intended to run multiple TF instances or use the GPU as a graphic redering device while running ROCm applications.

**Comment by sunway513** [comment]
Hi @seesturm , I'm closing this ticket for now. Please let us know if you have further questions.

**Comment by seesturm** [comment]
Hi @sunway513. Thanks for your analysis. But to be honest, I don't see my question answered. I'm still under the impression that if I'd report the same issue for an 8GB card I'd be "helped" by stating that learning should be performed on a card which is not connected to any monitor. If this really the case, then the answer to my question in the title is basically **No**.

The other thing which troubles me how obscure tensorflow-rocm is telling me that I'm running out of memory. I'd expect a clear message like "memory allocation failed". But what I'm getting is random hangs or memory access violations.

**Comment by sunway513** [comment]
Hi @seesturm , like what I can tell, just connecting my monitor to RX480 will consume 700MB+ device memory; and the amount of device memory allocated to graphics rendering would change dynamically based on the application running on your workstation. 
Given the default behavior of TF memory allocation policy (attempt to allocate all device memories at initialization), we recommend users not to use the GPU for graphics rendering and TF training concurrently to avoid such troubles you mentioned in the threads. 

I totally agree with your point that you should not see a vmem fault for such a case. The VMEM access fault should be related to the known defect in ROCm2.1 OpenCL tool chain for gfx803 ([link](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues?q=is%3Aissue+is%3Aopen+label%3Agfx803)). 
In my repro (without limiting TF to only allocate 1/3 of device memory), I did see the TF runtime reports memory allocation failure and rocblas handle creation failure like below:
```
2019-02-15 14:55:30.257537: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 184.25M (193202944 bytes) from device: hipError_t(1002)
2019-02-15 14:55:30.257557: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 165.83M (173882880 bytes) from device: hipError_t(1002)
2019-02-15 14:55:30.257577: E tensorflow/stream_executor/rocm/rocm_driver.cc:601] failed to allocate 149.24M (156494592 bytes) from device: hipError_t(1002)
2019-02-15 14:55:31.075238: E tensorflow/stream_executor/rocm/rocm_blas.cc:279] failed to create rocBLAS handle: rocblas_status_memory_error
2019-02-15 14:55:31.075759: E tensorflow/stream_executor/rocm/rocm_blas.cc:279] failed to create rocBLAS handle: rocblas_status_memory_error
2019-02-15 14:55:31.075971: E tensorflow/stream_executor/rocm/rocm_blas.cc:279] failed to create rocBLAS handle: rocblas_status_memory_error
```
However, we are not saying we don't support simultaneous graphical desktop usage. And in fact, for my moderate usage for web surfing while running TF training could work correctly on my RX480 GPU. In addition, if you can limit your TF instance memory allocation policy (e.g. allocate 1/3 of the device memory available), you should be able to run multiple TF instances as well. 
In the future, if you encounter a similar issue, we are here to help you sort those out.

**Comment by seesturm** [comment]
Update: I installed now a Radeon Vega 56 8GB card in my computer. With this card I can now run the cifar10 training together with prime95 and websurfing in parallel without any problem. This is based on two tests so the confidence is not yet 100%. But with my RX460 this scenario was failing all the time - it is at least a good indication that the situation has drastically improved.

Out of couriosity I'm planning to test fixes for the mentioned issues if they become available also on my RX460. Probably then I will only see a memory allocation error message, but this fix would give me further confidence in the tensorflow support.

Thanks for your assistance and responsiveness in the github issue tracker.

**Comment by sunway513** [comment]
Thanks @seesturm , great to know you see the improvements with Vega56!
Will keep you updated when we have an official fix for GFX803 vmem fault issue in the coming releases.
